pkg/streamingpromql/operators/aggregations/topkbottomk/range_query.go

```diff
@@ -0,0 +1,497 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/prometheus/prometheus/blob/main/promql/engine.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Prometheus Authors
+
+package topkbottomk
+
+import (
+	"container/heap"
+	"context"
+	"fmt"
+	"math"
+	"slices"
+	"sort"
+	"unsafe"
+
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/promql/parser/posrange"
+	"github.com/prometheus/prometheus/util/annotations"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/limiting"
+	"github.com/grafana/mimir/pkg/streamingpromql/operators/aggregations"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+	"github.com/grafana/mimir/pkg/util/pool"
+)
+
+// RangeQuery implements topk() and bottomk() for range queries.
+type RangeQuery struct {
+	Inner                    types.InstantVectorOperator
+	Param                    types.ScalarOperator
+	TimeRange                types.QueryTimeRange
+	Grouping                 []string // If this is a 'without' aggregation, New will ensure that this slice contains __name__.
+	Without                  bool
+	MemoryConsumptionTracker *limiting.MemoryConsumptionTracker
+	IsTopK                   bool // If false, this is operator is for bottomk().
+
+	expressionPosition posrange.PositionRange
+	limit              []int64 // Maximum number of values to return at each time step for each group
```

Comment:
`kLimits`

Reply:
I've renamed this to `k` in line with the "k" in "topk" and "bottomk".